### PR TITLE
[full-ci][tests-only] run acceptance tests using files for upload from the middleware

### PR DESCRIPTION
**Pull request by kiranparajuli589**
### Description:
use `filesForUpload` from the middleware repo to run web acceptance tests

### Related:
- see also the description in https://github.com/owncloud/owncloud-test-middleware/pull/71
- blocker in https://github.com/owncloud/web/pull/6188

### Motivation:
- prepare files/folders to upload for upload tests (with any client) from the middleware, so that we can assert the downloaded content with the local content easily.
- If we want to test an upload feature using a locally created file, then:
  - web-test have to create a file locally
  - save it
  - upload it to the selenium server
  - upload it to owncloud using the webUI
  - download the file using the webUI
  - downloaded content should match the local file content

- now the same tests may be needed for the desktop client or any other client. we use middleware to solve duplication issues and move steps like:
  - creating a local file
  - asserting downloaded content with local file
to the middleware so that every client can take benefit from this.

- also in the process, the folder used by middleware to manage filesForUploads can be shared with selenium (thanks to docker), and now we even do not have to upload these files to the selenium (since they are already synced from middleware dir.). then, we can make an upload test with ease.

### CODE REVIEW

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingPublic2</strong> https://drone.owncloud.com/owncloud/web/22038/38/1
:boom: The acceptance tests pipeline failed. The build has been cancelled.

**Comment by ownclouders** [comment]
Results for <strong>oCISSharingPublic2</strong> https://drone.owncloud.com/owncloud/web/22038/65/1
:boom: The acceptance tests pipeline failed. The build has been cancelled.

**Comment by ownclouders** [comment]
Results for <strong>e2e-tests oC10</strong> https://drone.owncloud.com/owncloud/web/22040/10/1
<details><summary>:boom: To see the trace, please open the link in the console ...</summary>

<p>


#### npx playwright show-trace https://cache.owncloud.com/owncloud/web/tracing/22040/2022-1-20-08-14-18-Alice.zip 

#### npx playwright show-trace https://cache.owncloud.com/owncloud/web/tracing/22040/2022-1-20-08-15-00-Alice.zip 

#### npx playwright show-trace https://cache.owncloud.com/owncloud/web/tracing/22040/2022-1-20-08-15-10-Brian.zip 


</p></details>

**Comment by ownclouders** [comment]
Results for <strong>oC10Move</strong> https://drone.owncloud.com/owncloud/web/22049/18/1
<details><summary>:boom: The acceptance tests failed on retry. Please find the screenshots inside ...</summary>

<p>


### webUIOperationsWithFolderShares-accessToSharesFolder_feature-L57.png
  


</p></details>

**Comment by ownclouders** [comment]
Results for <strong>oC10IntegrationApp2</strong> https://drone.owncloud.com/owncloud/web/22049/66/1
<details><summary>:boom: The acceptance tests failed on retry. Please find the screenshots inside ...</summary>

<p>


### webUIUpload-upload_feature-L34.png
  

### webUIUpload-upload_feature-L48.png
  


</p></details>

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingPublic2</strong> https://drone.owncloud.com/owncloud/web/22049/35/1
<details><summary>:boom: The acceptance tests failed on retry. Please find the screenshots inside ...</summary>

<p>


### webUISharingPublicDifferentRoles-shareByPublicLinkDifferentRoles_feature-L159.png
  

### webUISharingPublicDifferentRoles-shareByPublicLinkDifferentRoles_feature-L169.png
  

### webUISharingPublicDifferentRoles-shareByPublicLinkDifferentRoles_feature-L199.png
  

### webUISharingPublicDifferentRoles-shareByPublicLinkDifferentRoles_feature-L209.png
  


</p></details>

**Comment by ownclouders** [comment]
Results for <strong>oC10Files2</strong> https://drone.owncloud.com/owncloud/web/22049/14/1
<details><summary>:boom: The acceptance tests failed on retry. Please find the screenshots inside ...</summary>

<p>


### webUIUpload-upload_feature-L167.png
  

### webUIUpload-upload_feature-L34.png
  

### webUIUpload-upload_feature-L48.png
  


</p></details>

File: docs/testing-manual.md [hunk 1]
```diff
@@ -9,6 +9,24 @@ geekdocFilePath: testing-old.md
...
+
+- set up the [oCIS backend]({{< ref "backend-ocis.md" >}})
+  - if you are a Mac user, you need to start the server with additional environment variables: `STORAGE_HOME_DATA_SERVER_URL='http://host.docker.internal:9155/data' STORAGE_DATAGATEWAY_PUBLIC_URL='https://host.docker.internal:9200/data' STORAGE_USERS_DATA_SERVER_URL='http://host.docker.internal:9158/data' STORAGE_FRONTEND_PUBLIC_URL='https://host.docker.internal:9200' PROXY_ENABLE_BASIC_AUTH=true PROXY_OIDC_ISSUER='https://host.docker.internal:9200' IDP_INSECURE='true' IDP_IDENTIFIER_REGISTRATION_CONF='<web-path>/tests/acceptance/mac-identifier-registration.yml' IDP_ISS='https://host.docker.internal:9200' IDP_TLS='true'` (`<web-path>` needs to be replaced with the path of your local clone of ownCloud Web)
+
+## Setup the owncloud test Middleware
+To run the tests, we need to run the [owncloud-test-middleware](https://github.com/owncloud/owncloud-test-middleware). You can find the instructions on how to run the middleware server from its [github page](https://github.com/owncloud/owncloud-test-middleware#starting-the-server)
+
 ## Setup Selenium
```
Comment (kiranparajuli589) [inline docs/testing-manual.md:30]: `setup selenium` section is moved after the `server setup` & `middleware setup` as the selenium docker needs to run with the `filesForUpload` from the middleware code base.

**Comment by ownclouders** [comment]
Results for <strong>oC10Move</strong> https://drone.owncloud.com/owncloud/web/22071/18/1
<details><summary>:boom: The acceptance tests failed on retry. Please find the screenshots inside ...</summary>

<p>


### webUIMoveFilesFolders-moveFiles_feature-L20.png
  


</p></details>

File: docs/testing-manual.md [hunk 2]
```diff
@@ -9,6 +9,24 @@ geekdocFilePath: testing-old.md
...
 
+## Setup backend
+
+### ownCloud 10
+
+- set up the [ownCloud 10 backend]({{< ref "backend-oc10.md" >}})
+- clone and install the [testing app](http://github.com/owncloud/testing) into ownCloud
+
+### oCIS
+In order to run the acceptance tests you need to run oCIS using the oCIS storage driver. Also, you need to enable basic auth on the server with these environment variables.
```
Comment (saw-jan) [inline docs/testing-manual.md:20]: .
```suggestion
In order to run the acceptance tests, you need to run oCIS using the oCIS storage driver. Also, you need to enable basic auth on the server with these environment variables.
```

File: docs/testing-manual.md [hunk 3]
```diff
@@ -9,6 +9,24 @@ geekdocFilePath: testing-old.md
...
+
+`PROXY_ENABLE_BASIC_AUTH=true STORAGE_HOME_DRIVER=owncloud STORAGE_USERS_DRIVER=owncloud`
+
+- set up the [oCIS backend]({{< ref "backend-ocis.md" >}})
+  - if you are a Mac user, you need to start the server with additional environment variables: `STORAGE_HOME_DATA_SERVER_URL='http://host.docker.internal:9155/data' STORAGE_DATAGATEWAY_PUBLIC_URL='https://host.docker.internal:9200/data' STORAGE_USERS_DATA_SERVER_URL='http://host.docker.internal:9158/data' STORAGE_FRONTEND_PUBLIC_URL='https://host.docker.internal:9200' PROXY_ENABLE_BASIC_AUTH=true PROXY_OIDC_ISSUER='https://host.docker.internal:9200' IDP_INSECURE='true' IDP_IDENTIFIER_REGISTRATION_CONF='<web-path>/tests/acceptance/mac-identifier-registration.yml' IDP_ISS='https://host.docker.internal:9200' IDP_TLS='true'` (`<web-path>` needs to be replaced with the path of your local clone of ownCloud Web)
+
+## Setup the owncloud test Middleware
```
Comment (saw-jan) [inline docs/testing-manual.md:27]:
```suggestion
## Setup the ownCloud test Middleware
```

File: docker-compose.yml [hunk 4]
```diff
@@ -72,14 +72,17 @@ services:
...
-      - ./tests/acceptance/filesForUpload:/usr/src/app/filesForUpload
     environment:
       MIDDLEWARE_PORT: 3001
       MIDDLEWARE_HOST: "0.0.0.0"
       RUN_ON_OCIS: 'false'
       BACKEND_HOST: "http://host.docker.internal:8080"
+      REMOTE_UPLOAD_DIR: "/uploads"
 
   middleware-ocis:
     image: owncloud/owncloud-test-middleware:latest
```
Comment (SwikritiT) [inline docker-compose.yml:88]: The middleware version needs to be bumped as well

File: tests/acceptance/stepDefinitions/generalContext.js [hunk 5]
```diff
@@ -132,46 +119,6 @@ Then('no message should be displayed on the webUI', function () {
...
-  const expectedContent = fs.readFileSync(fullPathOfLocalFile, { encoding: 'utf-8' })
-  return assert.notStrictEqual(
-    actualContent,
-    expectedContent,
-    'asserting content of local file "' + fullPathOfLocalFile + '"'
-  )
-}
-
 Given(
   'the setting {string} of app {string} has been set to {string}',
```
Comment (SwikritiT) [inline tests/acceptance/stepDefinitions/generalContext.js:123]: These steps and some below are present in middleware. Why not use middleware's?

Comment (kiranparajuli589) [reply tests/acceptance/stepDefinitions/generalContext.js:123]: refactoring for this step is already done in https://github.com/owncloud/web/pull/6188 (which is blocked until this PR), so avoided here

File: docker-compose.yml (hunk 4 above)
Comment (kiranparajuli589) [reply docker-compose.yml:88]: should not be necessary as it is already using the `latest`

**Comment by ownclouders** [comment]
Results for <strong>oC10XGAPortrait1</strong> https://drone.owncloud.com/owncloud/web/22084/43/1
The following scenarios passed on retry:
- webUIPrivateLinks/accessingPrivateLinks.feature:9

**Comment by ownclouders** [comment]
Results for <strong>oC10Files2</strong> https://drone.owncloud.com/owncloud/web/22084/17/1
<details><summary>:boom: The acceptance tests failed on retry. Please find the screenshots inside ...</summary>

<p>


### webUIUpload-upload_feature-L73.png
  


</p></details>

**Comment by ownclouders** [comment]
Results for <strong>oC10Files2</strong> https://drone.owncloud.com/owncloud/web/22087/17/1
<details><summary>:boom: The acceptance tests failed on retry. Please find the screenshots inside ...</summary>

<p>


### webUIUpload-upload_feature-L73.png
  


</p></details>

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingAccept</strong> https://drone.owncloud.com/owncloud/web/22089/15/1
The following scenarios passed on retry:
- webUISharingAcceptSharesToRoot/acceptShares.feature:108

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingExternal</strong> https://drone.owncloud.com/owncloud/web/22089/40/1
The following scenarios passed on retry:
- webUISharingExternal/federationSharing.feature:106

File: dev/docker/middleware.entrypoint.sh [hunk 6]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env sh
+# cleanup filesForUpload dir before start
+rm -rf /uploads/*
+# populate shared volume with the files
+cp -r filesForUpload/* /uploads/
+# start the middleware service
+yarn start
```
Comment (individual-it) [inline dev/docker/middleware.entrypoint.sh:7]: I think we could move this to the middleware container itself

**Review by individual-it — CHANGES_REQUESTED** [review]
we should be able to get rid of the `uploadRemote()` function completely if we refactor 
https://github.com/owncloud/web/blob/9f5bd8e570f02b3be6f73826aae3b938f1bd4fbc/tests/acceptance/stepDefinitions/filesContext.js#L261-L265

This special upload folder can be created inside the global upload folder and through the volumes selenium will have access to it

**Comment by kiranparajuli589** [comment]
> we should be able to get rid of the `uploadRemote()` function completely if we refactor
> 
> https://github.com/owncloud/web/blob/9f5bd8e570f02b3be6f73826aae3b938f1bd4fbc/tests/acceptance/stepDefinitions/filesContext.js#L261-L265
> 
> This special upload folder can be created inside the global upload folder and through the volumes selenium will have access to it

since it is a `when` step implemented inside `web`, it has to know where to upload those created folders. so we've to use env `REMOTE_UPLOAD_DIR` to make the `web-tests` know about the remote upload directory.

we can remove the usage of `REMOTE_UPLOAD_DIR` completely if we create those folders in the given step (from middleware). open for your comments and suggestions.

File: dev/docker/middleware.entrypoint.sh (hunk 6 above)
Comment (kiranparajuli589) [reply dev/docker/middleware.entrypoint.sh:7]: the `/uploads` folder directly depends on the compose file living in `web` and volumes used by the container. if we change the volume path in the `docker-compose` file, we've to come to the middleware repo just to fix the script. should we still move this?

Comment (kiranparajuli589) [reply dev/docker/middleware.entrypoint.sh:7]: the `/uploads` folder directly depends on the compose file living in `web` and volumes used by the container. if we change the volume path in the `docker-compose` file, we've to come to the middleware repo just to fix the script. should we still move this?

**Comment by ownclouders** [comment]
Results for <strong>oC10IntegrationApp2</strong> https://drone.owncloud.com/owncloud/web/22163/69/1
<details><summary>:boom: The acceptance tests failed on retry. Please find the screenshots inside ...</summary>

<p>


### webUIUpload-upload_feature-L22.png
  


</p></details>

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingPublic1</strong> https://drone.owncloud.com/owncloud/web/22163/37/1
The following scenarios passed on retry:
- webUISharingPublicBasic/publicLinkPublicActions.feature:18

**Comment by ownclouders** [comment]
Results for <strong>oCISSharingAndUpload</strong> https://drone.owncloud.com/owncloud/web/22163/63/1
<details><summary>:boom: The acceptance tests failed on retry. Please find the screenshots inside ...</summary>

<p>


### webUIUpload-upload_feature-L22.png
  


</p></details>

**Comment by ownclouders** [comment]
Results for <strong>oC10Files2</strong> https://drone.owncloud.com/owncloud/web/22163/17/1
<details><summary>:boom: The acceptance tests failed on retry. Please find the screenshots inside ...</summary>

<p>


### webUIUpload-upload_feature-L22.png
  


</p></details>

**Review by individual-it — APPROVED** [review]
LGTM

CC @pascalwengerter & @kulmann

**Comment by individual-it** [comment]
grafik
 lines shifted?

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingAccept</strong> https://drone.owncloud.com/owncloud/web/22176/15/1
The following scenarios passed on retry:
- webUISharingAcceptSharesToRoot/acceptShares.feature:108

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingAccept</strong> https://drone.owncloud.com/owncloud/web/22182/15/1
The following scenarios passed on retry:
- webUISharingAcceptSharesToRoot/acceptShares.feature:108

**Comment by ownclouders** [comment]
Results for <strong>oCISTrashbinUploadMoveJourney</strong> https://drone.owncloud.com/owncloud/web/22182/66/1
The following scenarios passed on retry:
- webUIUserJourney/journey1.feature:11

File: tests/acceptance/features/webUIUpload/upload.feature [hunk 7]
```diff
@@ -16,19 +16,21 @@ Feature: File Upload
     When the user uploads file "new-lorem.txt" using the webUI
     Then no message should be displayed on the webUI
     And file "new-lorem.txt" should be listed on the webUI
-    And as "Alice" the content of "new-lorem.txt" should be the same as the content of local file "new-lorem.txt"
+    And as "Alice" the content of "new-lorem.txt" in the server should be the same as the content of local file "new-lorem.txt"
 
   @smokeTest @ocisSmokeTest
   Scenario: simple upload of a folder that does not exist before
-    When the user uploads a folder containing the following files in separate sub-folders using the webUI:
+    Given a folder "CUSTOM" has been created with the following files in separate sub-folders in the server
```
Comment (phil-davis) [inline tests/acceptance/features/webUIUpload/upload.feature:23]: I don't understand why the words "in separate sub-folders" are in this step. All the files are in the same folder, in this case a folder called "CUSTOM".

"in the server" is a bit odd here. The files seem to actually be created "by the middleware" and are then visible to the test-runner to use as the source of upload.

But we could refactor these words in a separate PR later.

File: tests/acceptance/features/webUIUpload/upload.feature [hunk 8]
```diff
@@ -63,35 +65,35 @@ Feature: File Upload
 
   @smokeTest @ocisSmokeTest
   Scenario: uploading a big file (when chunking is implemented this upload should be chunked)
-    Given a file with the size of "30000000" bytes and the name "big-video.mp4" has been created locally
+    Given a file with the size of "30000000" bytes and the name "big-video.mp4" has been created locally in the server
```
Comment (phil-davis) [inline tests/acceptance/features/webUIUpload/upload.feature:68]: This step also reds a bit strange - "created locally in the server"

For steps that are done "in the middleware" but where the middleware does not do anything "in/on/to the server", we could improve the words. Maybe we can end steps with "in the middleware"?

So this step would end with "has been created locally in the middleware".

And we can easily have the test code match steps that end in "in the middleware"

File: tests/acceptance/features/webUIUpload/upload.feature [hunk 9]
```diff
@@ -63,35 +65,35 @@ Feature: File Upload
...
     When the user uploads a created file "big-video.mp4" using the webUI
     Then no message should be displayed on the webUI
     And file "big-video.mp4" should be listed on the webUI
-    And as "Alice" the content of "big-video.mp4" should be the same as the content of local file "big-video.mp4"
+    And as "Alice" the content of "big-video.mp4" should be the same as the content of local file "big-video.mp4" in the server
```
Comment (phil-davis) [inline tests/acceptance/features/webUIUpload/upload.feature:72]:
```suggestion
    And as "Alice" the content of "big-video.mp4" should be the same as the content of local file "big-video.mp4" in the middleware
```

I can understand "in the middleware" more easily. This step compares the "local" file that the middleware created in its local file system with the file that is stored in the server.

Maybe even say:
```
    And as "Alice" the content of "big-video.mp4" on the server should be the same as the content of local file "big-video.mp4" in the middleware
```

**Review by phil-davis — COMMENTED** [review]
Just some thoughts about some of the "in the server" steps.
Should we try to improve the words now, or get this PR done, and make a spearte issue and PR to refactor some of the words?

File: tests/acceptance/features/webUIUpload/upload.feature (hunk 8 above)
Comment (kiranparajuli589) [reply tests/acceptance/features/webUIUpload/upload.feature:68]: yes, @phil-davis, the server is not involved at all, so refactorings should be made. I've created an issue for this https://github.com/owncloud/owncloud-test-middleware/issues/96

File: tests/acceptance/features/webUIUpload/upload.feature (hunk 9 above)
Comment (kiranparajuli589) [reply tests/acceptance/features/webUIUpload/upload.feature:72]: issue created for this, we can do this after this PR

File: tests/acceptance/features/webUIUpload/upload.feature (hunk 7 above)
Comment (kiranparajuli589) [reply tests/acceptance/features/webUIUpload/upload.feature:23]: @phil-davis the given step has to be fixed in the middleware. It was working the same as the step description before but using session ids as folder names. We can store every file provided to some sub-folder like `sub1`, `sub2` according to the order, or even provide table data with sub-folder names.

**Comment by phil-davis** [comment]
@pascalwengerter & @kulmann

Are you OK for this to be merged. And @kiranparajuli589 will refactor some step words in a next PR for issue https://github.com/owncloud/owncloud-test-middleware/issues/96

**Review by phil-davis — APPROVED** [review]

**Comment by kiranparajuli589** [comment]
Also, after this PR, e2e tests can use their own small set of files for upload within, and `tests/acceptance/filesForUpload` can be completely removed.

File: tests/acceptance/features/webUIUpload/upload.feature (hunk 7 above)
Comment (phil-davis) [reply tests/acceptance/features/webUIUpload/upload.feature:23]: At the moment that step seems to create each file directly in the specified folder, there are no sub-folders created.

"or even provide table data with sub-folder names" - that sounds good.

We can refactor in the middleware in the future to have a step like:
```
Given a folder "CUSTOM" has been created with the following files and folders in the middleware
  | folder | file |
  |  | topFile.txt |
  | sub1 | someFile.txt |
  | sub1 | anotherFile.txt |
  | sub2/sub3 | file.txt |
  | sub4 |  |
  | sub5/sub6/sub7 |  |
```

If the folder column is empty, then the file goes directly in the main folder.
If the file column is empty, then the folder(s) are created, with the bottom folder empty
If the folder column has a multiple-folder path then create that whole folder tree.

IMO that will make a flexible step for creating whole file-folder trees in the middleware that can be used in a scenario.

**Comment by pascalwengerter** [comment]
> Also, after this PR, the e2e test can use its own small set of files for upload within, and `tests/acceptance/filesForUpload` can be completely removed.

Would you mind actually moving the files into the `e2e` directory and deleting them in this PR? :) would be cleaner from my POV

**Comment by kiranparajuli589** [comment]
@pascalwengerter, files moved inside `e2e` root. please review again.

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingAccept</strong> https://drone.owncloud.com/owncloud/web/22236/15/1
The following scenarios passed on retry:
- webUISharingAcceptSharesToRoot/acceptShares.feature:108

**Comment by ownclouders** [comment]
Results for <strong>oC10Move</strong> https://drone.owncloud.com/owncloud/web/22236/21/1
The following scenarios passed on retry:
- webUIOperationsWithFolderShares/accessToSharesFolder.feature:84

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingSharingInd</strong> https://drone.owncloud.com/owncloud/web/22236/33/1
The following scenarios passed on retry:
- webUISharingInternalGroupsSharingIndicator/shareWithGroups.feature:136

**Review by pascalwengerter — APPROVED** [review]
LGTM, thanks @kiranparajuli589! One red pipeline, not sure if flaky test or rebase necessary?

**Comment by kiranparajuli589** [comment]
hmm, _webUIMoveFilesFolders/moveFolders.feature:20_ failed on `oc10iPhone` suite. Rebase was not necessary so, I've restarted the CI, let's see if it passes this time.

**Review by phil-davis — APPROVED** [review]
LGTM - e2e and acceptance tests all passed. Let's merge.

**Pull request merged**